lib/buffer-writer.js

```diff
@@ -0,0 +1,208 @@
+    const headerBytes = CBOR.encode({ version: 1, roots })
+    const varintBytes = varint.encode(headerBytes.length)
+
+    const headerByteLength = varintBytes.length + headerBytes.byteLength
+    const offset = headerCapacity - headerByteLength
+
+    if (offset === 0) {
+      writeHeader(varintBytes, headerBytes, bytes)
+      return bytes.subarray(0, byteOffset)
+    // If header was misestimated, yet buffer has capacity to fit header and
+    // written blocks
+    } else if (byteOffset + offset < bytes.byteLength) {
+      // If we `align: true` is passed we will align header & blocks as needed.
+      if (align) {
+        this.byteOffset += offset
+        // Move encoded blocks to a new offset. Please note that we may have
+        // underestimated header so this needs to happen before we write header.
+        bytes.set(bytes.subarray(headerCapacity, byteOffset), headerByteLength)
+
+        writeHeader(varintBytes, headerBytes, bytes)
+        return bytes.subarray(0, this.byteOffset)
+      } else {
+        throw new RangeError(
+
+          `Header size was ${offset > 0 ? `underestimated by ${offset}` : `overestimated by ${-1 * offset}`}
+You can use close({ align: true }) to align header and blocks as needed`)
+      }
+    } else {
+      throw new RangeError(`Header size was underestimated by ${-1 * offset} bytes and there is not enough space in the buffer now`)
+    }
+  }
+}
+
+/**
+ *
+ * @param {number[]} varint
+ * @param {Uint8Array} header
+ * @param {Uint8Array} destination
+ * @param {number} offset
+ */
+const writeHeader = (varint, header, destination, offset = 0) => {
+  destination.set(varint, offset)
+  destination.set(header, offset + varint.length)
+}
+
+/**
+ * Attempts to estimate header size given number of roots it will have assuming
+ * they're CIDv1 in with sha256 digest. Optionally it takes number of bytes
+ * to be allocated for all the roots, in case different hashing or CID version
+ * is used.
+ *
+ * Note: Returned value is just an estimate which can be inaccurate where large
+ * number of CIDs is passed or if they are of various sizes.
+ *
+ * @param {number} count - Number of roots
+ * @param {number} [rootsByteLength] - Total byteLength of all roots
+ */
+export const estimateHeaderSize = (
+  count,
+  rootsByteLength = count * DEFAULT_CID_SIZE
+) => {
+  const lengthSize = arrayLengthEncodeSize(count)
+  const rootsSize = rootsByteLength + count * ROOT_EXTRA_SIZE
+  const headerSize = EMPTY_HEADER_SIZE + lengthSize + rootsSize
+  const varintSize = varint.encodingLength(headerSize)
+  return varintSize + headerSize
+}
+
+/**
+ *
+ * @param {number} length
+ * @returns
+ */
+const arrayLengthEncodeSize = length =>
+  length < 24
+    ? 1
+    : length < 256
+      ? 2
+      : length < 65536
+        ? 3
+        : CBOR.encode(length).length
+
+/**
+ * @param {CID[]} cids
+ */
+const totalByteLength = (cids) => {
+  let total = 0
+  for (const cid of cids) {
+    total += cid.bytes.byteLength
```

Comment:
We could fix this too now you have the nice `arrayLengthEncodeSize()` which will work for CIDs too; I think if you subtract `2` from `ROOT_EXTRA_SIZE` (the byte array is >24<256 so it requires two bytes cbor prelude `<bytes><length>`, the extra bits in `ROOT_EXTRA_SIZE` are to do with tags, mainly)  and then add it back in by adding `+ arrayLengthEncodeSize(cid.length)` here then you get a more accurate CID length. The problem with doing that is that your API allows for an optional `rootsByteLength`, but that could also probably be fixed by adding those `2` to the `DEFAULT_CID_SIZE`.

If you add `bafkqaaia` and `bafkqbbacaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa` to your tests then you should see the impact of the bad sizes here.

Identity CIDs are the most likely way we're going to end up here--it would be silly to use one as a root but it's allowed and it's not out of the question that someone finds a use-case for it.